Design/Fit functions in python.md

```diff
@@ -0,0 +1,179 @@
+```
+  g = Gaussian()
+```
+A call without passing in any initialisation arguments should create a function with default parameters and attributes.
+Initial values of parameters and attributes can be given via keyword arguments, for example
+```
+  p = Polynomial(attributes={'n': 3}, parameters={'A0': 1, 'A1': 2, 'A3': 3, 'A4': 4})
+```
+Parameter and attribute names can also be used as the keys:
+```
+  g = Gaussian(Height=1, Sigma=0.1)
+  
+  p = Polynomial(n=3, A0=1, A1=2, A3=3, A4=4)
+```
+The implementation must check each keyword argument if it's a parameter or an attribute by calling `IFunction::hasAttribute(name)`. All passed attributes must be set before setting parameters. There is need to be a provision for setting attributes in a particular order in case it is essential for a fitting function. This can be done for example by passing a list of name/value pairs (tuples or lists) to the `attributes` keyword argument.
+
+### Construction of composite functions
+
+Composite functions can be created in a similar way, by callig a constructor. It should be able to accept the same types of arguments as a simple function but also it must have the `functions` keyword argument that takes a list of member functions. For example:
+```
+  c = CompositeFunction(functions=[Gaussian(PeakCentre=1), Gaussian(PeakCentre=2)])
+```
+Another way of setting member functions is via positional arguments of the condtructor:
+```
+  c = CompositeFunction(Gaussian(PeakCentre=1), Gaussian(PeakCentre=2))
+```
+`FitFunctionWrapper` class will override the `__add__()` method to allow construction of composite functions from a sum of member functions:
+```
+  c = LinearBackground() + Gaussian(PeakCentre=1) + Gaussian(PeakCentre=2)
+```
+`ProductFunction` can be constructed from a product of member functions:
+```
+  p = ExpDecay() * UserFunction(Formula='sin(w*x)', w=1)
+```
+
+### Multi-domain functions
+
+A [`MultiDomainFunction`](https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/MultiDomainFunction.h) needs a custom constructor which sets domain indices to its members.
+
+## Updating functions
+
+### Getting and setting parameters and attributes
+
+`FitFunctionWrapper` will implement `__getitem__(self, name)` and `__setitem__(self, name, value)` methods to access parameters and attributes in a dictionary-like style (`name` is a string):
+```
+  sigma = gaussian['Sigma']
+  
+  comp_func['f1.A0'] = 0.5
+```
+
+Parameters of members of composite functions can be accessed via their wrapper objects. For example:
+```
+  bk = LinearBackground()
+  peak = Lorentzian()
+  spectrum = bk + peak
+  ...
+  bk['A0'] = 1
+  peak['FWHM'] = 0.123
+  assert spectrum['f0.A0'] == 1
+  assert spectrum['f1.FWHM'] == 0.123
+```
+*Question: what to do if a function becomes a member of two or more composite functions?*
```

Comment:
I think setting some kind of flag is possibly overkill.

The question here is what kind of behaviour is least confusing to users:

1. take a copy of the function as it's added to a composite so that all subsequent changes only affect that object or
2. leave it has you have here and be clear in the documentation that altering objects after are added to the composite will continue to affect it.

I'm tempted to say that the copy approach will be less error-prone and easier to explain.